File: code/design-patterns/behavioral/chain-of-responsibility/go/chain_of_responsibility_usage.go
package main

import "fmt"

// Main function
func main() {
	// Create the departments
	reception := &Reception{}
	doctor := &Doctor{}
	medicalExaminationRoom := &MedicalExaminationRoom{}
	cashier := &Cashier{}

	// Set up the chain of responsibility
	reception.SetNext(doctor).SetNext(medicalExaminationRoom).SetNext(cashier)

	// Create a patient
	patient := &Patient{
		Insurance: true,
		Broke:     false,
		Log:       "",
	}

	// Handle the patient through the chain of responsibility
	reception.Handle(patient)
	fmt.Println("Patient Log:", patient.Log)
	if patient.Log == "Reception -> Doctor -> MedicalExaminationRoom -> Cashier -> " {
		fmt.Println("Test passed for patient 1")
	} else {
		fmt.Println("Test failed for patient 1")
	}

	// Create another patient without insurance
	patient2 := &Patient{
		Insurance: false,
		Broke:     false,
		Log:       "",
	}

	// Handle the second patient through the chain of responsibility
	reception.Handle(patient2)
	fmt.Println("Patient2 Log:", patient2.Log)
	if patient2.Log == "" {
		fmt.Println("Test passed for patient 2")
	} else {
		fmt.Println("Test failed for patient 2")
	}
}

File: code/design-patterns/behavioral/chain-of-responsibility/python/chain_of_responsibility_usage.py
from chain_of_responsibility import Reception, Doctor, MedicalExaminationRoom, Cashier

reception = Reception()
doctor = Doctor()
medical_examination_room = MedicalExaminationRoom()
cashier = Cashier()

reception.set_next(doctor).set_next(medical_examination_room).set_next(cashier)

patient = {"insurance": True, "broke": False, "log": ""}
reception.handle(patient)
assert (
    patient.get("log") == "Reception -> Doctor -> MedicalExaminationRoom -> Cashier -> "
)

patient2 = {"insurance": False, "broke": False, "log": ""}
reception.handle(patient2)
print(patient2.get("log"))
assert patient2.get("log") == ""
